### DISCUSSION TITLE
Prevent Table Header Sorting on Full Header Click — Sorting Should Only Trigger on Sort Icon Click

### DISCUSSION THREAD

**Discussion by Hashim1995** [ant-design/ant-design#51358] [Q&A]
I’m using the Ant Design Table component, and I’ve encountered an issue where sorting is triggered when a user clicks anywhere within the table header cell (th), even if they are interacting with non-sorting elements inside the header. Ideally, sorting should only be triggered when the user explicitly clicks the sorting icon.

Currently, even if a header contains custom elements like resize handles or custom titles, clicking anywhere inside the header cell initiates sorting. This behavior makes it challenging to provide custom header interactions without accidentally sorting the column.

Steps to Reproduce:

Create a table with sortable columns.
Add a custom element (e.g., a resizable column handle or custom title) inside the header cell.
Click anywhere inside the header cell that is not the sort icon.
Expected Behavior: Sorting should only be triggered when the user clicks directly on the sorting icon (.ant-table-column-sorters svg). Other interactions inside the header cell should not trigger sorting.

Actual Behavior: Clicking anywhere inside the header cell triggers sorting, regardless of whether the user clicked the sort icon or not.

Suggested Solution: It would be helpful if sorting could be limited to clicks directly on the sorting icon, or if there were a more explicit way to differentiate between sorting and other interactions in the table header.

Versions:

Ant Design: ^5.17.4
React: ^18.3.1
Browser: Chrome

Thank you for your hard work on the Ant Design project! I believe improving this behavior will allow for more flexible and interactive table headers.

**Comment by petercat-assistant** [comment]
Thank you for sharing the details of the issue you're experiencing with the Ant Design Table component. It seems like the current behavior of the table header sorting is not aligning with your needs, especially when custom elements are involved in the header.

To address this, you can consider the following workaround until an official solution is provided:

1. **Custom Click Handling**: You can override the default click behavior on the table header cells by adding a custom click event handler. This handler should check if the click event originated from the sorting icon and only then trigger the sorting logic.

2. **Event Delegation**: Use event delegation to determine the exact target of the click event. You can inspect the event target and decide whether to proceed with sorting based on whether the click was on the sorting icon.

Here's a basic example of how you might implement this:

```jsx
import React from 'react';
import { Table } from 'antd';

const CustomTable = () => {
  const columns = [
    {
      title: 'Name',
      dataIndex: 'name',
      sorter: true,
      onHeaderCell: (column) => ({
        onClick: (e) => {
          if (e.target.closest('.ant-table-column-sorters svg')) {
            // Trigger sorting logic here
            console.log('Sorting triggered');
          } else {
            // Prevent sorting
            e.stopPropagation();
            console.log('Sorting prevented');
          }
        },
      }),
    },
    // Add more columns as needed
  ];

  const data = [
    { key: '1', name: 'John Brown' },
    { key: '2', name: 'Jim Green' },
    // Add more data as needed
  ];

  return <Table columns={columns} dataSource={data} />;
};

export default CustomTable;
```

This code snippet demonstrates how to intercept the click event on the header cell and check if the click was on the sorting icon. If not, it prevents the default sorting behavior.

For further assistance, please describe your question in the comments and @petercat-assistant to start a conversation with me.

**Reply by Hashim1995** [reply]
I had already tried a similar approach by using onHeaderCell to stop the propagation of the click event on the header cell. However, this method does not prevent sorting behavior entirely, as clicking anywhere within the table header (outside of the sorting icon) still triggers the sorting mechanism.

**Comment by nouman2075** [comment]
@Hashim1995 Try this

```javascript
import { Table } from 'antd';
import type { TableColumnsType, TableProps } from 'antd';
import type React from 'react';
import type { ComponentPropsWithoutRef } from 'react';

interface DataType {
	key: React.Key;
	name: string;
	age: number;
}

const sortIconClassName = 'custom-icon';

const HeaderCell = (props: ComponentPropsWithoutRef<'th'>) => {
	return (
		<th
			{...props}
			onClick={(e) => {
				const currentTarget = e.currentTarget as HTMLElement;
				const sortIconNode = currentTarget.querySelector(`.${sortIconClassName}`);
				if (sortIconNode?.contains(e.target as HTMLElement)) {
					props.onClick?.(e);
				}
			}}
		/>
	);
};

const components: TableProps['components'] = {
	header: {
		cell: HeaderCell
	}
};

const data: Array<DataType> = [
	{ key: '1', name: 'John Brown', age: 20 },
	{ key: '2', name: 'Jim Green', age: 30 },
	{ key: '3', name: 'Joe Black', age: 40 },
	{ key: '4', name: 'Jim Red', age: 50 }
];

const App: React.FC = () => {
	const columns: TableColumnsType<DataType> = [
		{
			title: 'Name',
			dataIndex: 'name'
		},
		{
			title: 'Age',
			dataIndex: 'age',
			sorter: { compare: (a, b) => a.age - b.age },
			showSorterTooltip: { target: 'sorter-icon', className: sortIconClassName }
		}
	];

	return <Table<DataType> columns={columns} dataSource={data} components={components} />;
};

export default App;
```

**Comment by rghaemifar** [comment]
I searched the whole internet but found nothing
So I came up with my solution:

- Removed the sort event handlers and sort styles from the whole th tag
- Injected the sort event handlers into the sorter icon instead

Here is my Antd table:
```tsx
    <Table
      {...props}
      components={{
        header: {
          cell: Th,
        },
      }}
    />
```

And this is the custom Th component:
```tsx
import {
  Children,
  cloneElement,
  ComponentPropsWithoutRef,
  HTMLAttributes,
  isValidElement,
  ReactElement,
  ReactNode,
  useMemo,
} from 'react'

type Span = ReactElement<HTMLAttributes<HTMLSpanElement>>

function Th(props: ComponentPropsWithoutRef<'th'>) {
  // Remove sort event handlers from the whole cell
  const { className = '', onClick, onKeyDown, ...rest } = props
  const { 'aria-sort': areaSort } = rest

  const [classNameWithoutSorter, hasSorter] = useMemo(() => {
    const newClassName = className?.replace('ant-table-column-has-sorters', '')
    const hasSorter = newClassName.length !== className.length
    return [newClassName, hasSorter]
  }, [className])

  const recursivelyCloneChildren = (children: ReactNode): ReactNode => {
    return Children.map(children, (child) => {
      if (!isValidElement(child)) return child

      if (
        child.type === 'span' &&
        child.props?.className?.includes('ant-table-column-sorter')
      ) {
        // inject sort event handlers into sorter icon span
        return cloneElement(child as Span, {
          onClick,
          onKeyDown,
          style: {
            cursor: 'pointer',
            display: 'flex',
          },,
        })
      }

      if (child.props.children)
        return cloneElement(child as ReactElement, {
          children: recursivelyCloneChildren(child.props.children),
        })

      return child
    })
  }

  const modifiedChildren = useMemo(
    () => (hasSorter ? recursivelyCloneChildren(rest.children) : undefined),
    [areaSort]
  )

  if (!hasSorter) return <th {...props} />
  return (
    <th {...rest} className={classNameWithoutSorter}>
      {modifiedChildren}
    </th>
  )
}

export default Th
```

I tried to optimize it and limit extra re-calculations using useMemo. However had to use React.cloneElement recursively (4 level) to be able to make new elements having sort event handlers in sorter react node. Because react nodes are read only and you cannot inject properties into them.
